docxcompose/properties.py

```diff
@@ -198,96 +199,215 @@ def set_properties(self, properties):
+    def replace_field_with_value(self):
+        """ Removes the field from the document, replacing it with
+        its value.
+        """
+        raise NotImplementedError()
+
+    def _get_fieldname_string(self):
+        raise NotImplementedError()
+
+    def _parse_fieldname(self):
+        match = self.fieldname_search_expr.search(self._get_fieldname_string())
+        if match is None:
+            return None
+        return match.groups()[0]
+
+
+class SimpleField(FieldBase):
+    """ Represents a simple field, i.e. <w:fldSimple> node in the
+    document.xml, its body containing the value of the field.
+    self.node here is the <w:fldSimple> node.
+    """
+
+    attr_name = "{{{}}}instr".format(NS["w"])
+
+    def _get_fieldname_string(self):
+        return self.node.attrib[self.attr_name]
+
+    def update(self, value):
+        text = xpath(self.node, './/w:t')
+        if text:
+            text[0].text = self._format_value(value)
+
+    def replace_field_with_value(self):
+        parent = self.node.getparent()
+        index = list(parent).index(self.node)
+        w_r = deepcopy(self.node[0])
+        parent.remove(self.node)
+        parent.insert(index, w_r)
+
+
+class ComplexField(FieldBase):
+    """ Represents a complex field, i.e. a several <w:r> nodes delimited by runs
+    containing <w:fldChar w:fldCharType="begin"/> and <w:fldChar w:fldCharType="end"/>.
+    In these fields, the actual value is stored in <w:r> nodes that come after a
+    <w:r><w:fldChar w:fldCharType="separate"/></w:r> node.
+    """
+
+    XPATH_PRECEDING_BEGINS = "./preceding-sibling::w:r/w:fldChar[@w:fldCharType=\"begin\"]/.."
+    XPATH_FOLLOWING_ENDS = "./following-sibling::w:r/w:fldChar[@w:fldCharType=\"end\"]/.."
+    XPATH_FOLLOWING_SEPARATES = "./following-sibling::w:r/w:fldChar[@w:fldCharType=\"separate\"]/.."
+
+    def __init__(self, field_node):
+        super(ComplexField, self).__init__(field_node)
+        # run and paragraph containing the field
+        self.w_r = self.node.getparent()
+        self.w_p = self.w_r.getparent()
+
+    def _get_fieldname_string(self):
+        return self.node.text
+
+    @property
+    def begin_run(self):
+        return xpath(self.w_r, self.XPATH_PRECEDING_BEGINS)[-1]
+
+    @property
+    def end_run(self):
+        if not hasattr(self, "_end_run"):
+            self._end_run = xpath(self.w_r, self.XPATH_FOLLOWING_ENDS)[0]
+        return self._end_run
+
+    @property
+    def separate_run(self):
+        separate = xpath(self.w_r, self.XPATH_FOLLOWING_SEPARATES)[0]
+
+        # The ooxml format standard says that the separate node is optional, but
+        # we do not have an implementation that supports a missing separate node.
+        # so we assert that the separte node we found is in our complex field
+        if not self.w_p.index(separate) < self.w_p.index(self.end_run):
+            msg = "Complex field without separate node is not supported"
+            raise NotImplementedError(msg)
```

Comment:
`NotImplmenetedError` is not the correct exception to use here, it should be used for abstract base classes only as per documentation https://docs.python.org/2.7/library/exceptions.html#exceptions.NotImplementedError. I would recommend to implement a custom `Exception` subclass for us. That also allows for easier searching/aggregating in `sentry`, i'd expect this might be raised once in a while 😉.

Reply:
Ok, did not know that. Thanks!